src/filldist.jl

```diff
@@ -30,21 +30,18 @@ end
 function _flat_logpdf(dist, x)
     if toflatten(dist)
         f, args = flatten(dist)
-        return sum(f.(args..., x))
+        return mapreduce(xi -> f(args..., xi), +, x)
```

Comment:
Let's try, I hope it would not break Tracker.

Reply:
Alas, this triggers `Tracker` method ambiguity
```
Multivariate distributions: Error During Test at C:\Users\astukalov\.julia\dev\DistributionsAD\test\ad\utils.jl:357
  Test threw exception
  Expression: ≈(Tracker.data((Tracker.gradient(f, x))[1]), finitediff, rtol = rtol, atol = atol)
  MethodError: -(::ForwardDiff.Dual{Nothing, Float64, 1}, ::Tracker.TrackedReal{Float64}) is ambiguous. Candidates:
    -(a::Real, b::Tracker.TrackedReal) in Tracker at C:\Users\astukalov\.julia\packages\Tracker\9xWLl\src\lib\real.jl:96
    -(x::ForwardDiff.Dual{Tx}, y::Real) where Tx in ForwardDiff at C:\Users\astukalov\.julia\packages\ForwardDiff\wAaVJ\src\dual.jl:144
  Possible fix, define
    -(::ForwardDiff.Dual{Tx}, ::Tracker.TrackedReal) where Tx
  Stacktrace:
    [1] zval(μ::Tracker.TrackedReal{Float64}, σ::Tracker.TrackedReal{Float64}, x::ForwardDiff.Dual{Nothing, Float64, 1})
      @ StatsFuns C:\Users\astukalov\.julia\packages\StatsFuns\vxSkw\src\distrs\norm.jl:10
    [2] normlogpdf(μ::Tracker.TrackedReal{Float64}, σ::Tracker.TrackedReal{Float64}, x::ForwardDiff.Dual{Nothing, Float64, 1})
      @ StatsFuns C:\Users\astukalov\.julia\packages\StatsFuns\vxSkw\src\distrs\norm.jl:39
    [3] logpdf
      @ C:\Users\astukalov\.julia\packages\Distributions\QLJcf\src\univariate\continuous\logitnormal.jl:126 [inlined]
    [4] #200
      @ .\none:0 [inlined]
    [5] #21
      @ C:\Users\astukalov\.julia\dev\DistributionsAD\src\filldist.jl:33 [inlined]
    [6] partial(f::DistributionsAD.var"#21#22"{Tuple{Tracker.TrackedReal{Float64}, Tracker.TrackedReal{Float64}}, DistributionsAD.var"#200#201"}, Δ::Tracker.TrackedReal{Float64}, i::Int64, args::Float64)
      @ Tracker C:\Users\astukalov\.julia\packages\Tracker\9xWLl\src\lib\array.jl:573
    [7] _broadcast_getindex_evalf
      @ .\broadcast.jl:670 [inlined]
    [8] _broadcast_getindex
      @ .\broadcast.jl:643 [inlined]
    [9] getindex
      @ .\broadcast.jl:597 [inlined]
   [10] copy
      @ .\broadcast.jl:899 [inlined]
   [11] materialize(bc::Base.Broadcast.Broadcasted{Base.Broadcast.DefaultArrayStyle{1}, Nothing, typeof(Tracker.partial), Tuple{Base.RefValue{DistributionsAD.var"#21#22"{Tuple{Tracker.TrackedReal{Float64}, Tracker.TrackedReal{Float64}}, DistributionsAD.var"#200#201"}}, Vector{Tracker.TrackedReal{Float64}}, Int64, Vector{Float64}}})
      @ Base.Broadcast .\broadcast.jl:860
   [12] broadcast(::typeof(Tracker.partial), ::Base.RefValue{DistributionsAD.var"#21#22"{Tuple{Tracker.TrackedReal{Float64}, Tracker.TrackedReal{Float64}}, DistributionsAD.var"#200#201"}}, ::Vector{Tracker.TrackedReal{Float64}}, ::Int64, ::Vararg{Any})
      @ Base.Broadcast .\broadcast.jl:798
   [13] ∇broadcast
      @ C:\Users\astukalov\.julia\packages\Tracker\9xWLl\src\lib\array.jl:577 [inlined]
   [14] copy(bc::Base.Broadcast.Broadcasted{Tracker.TrackedStyle, Tuple{Base.OneTo{Int64}}, typeof(Tracker.partial), Tuple{Base.RefValue{DistributionsAD.var"#21#22"{Tuple{Tracker.TrackedReal{Float64}, Tracker.TrackedReal{Float64}}, DistributionsAD.var"#200#201"}}, Vector{Tracker.TrackedReal{Float64}}, Int64, TrackedArray{…,Vector{Float64}}}})
      @ Tracker C:\Users\astukalov\.julia\packages\Tracker\9xWLl\src\lib\array.jl:607
   [15] materialize
      @ .\broadcast.jl:860 [inlined]
   [16] (::Tracker.var"#621#623"{Vector{Tracker.TrackedReal{Float64}}, DistributionsAD.var"#21#22"{Tuple{Tracker.TrackedReal{Float64}, Tracker.TrackedReal{Float64}}, DistributionsAD.var"#200#201"}, Tuple{TrackedArray{…,Vector{Float64}}}})(i::Int64)
      @ Tracker C:\Users\astukalov\.julia\packages\Tracker\9xWLl\src\lib\array.jl:580
```
I guess it has to be fixed on the `Tracker` side. Also looks like it still generates an array.

Reply:
> I guess it has to be fixed on the `Tracker` side. Also looks like it still generates an array.

Well, Tracker defines `sum(f, xs::TrackedArray)` as `sum(f.(xs))`: https://github.com/FluxML/Tracker.jl/blob/84ff74daaa250dadb424d251c66d64fa64ade819/src/lib/array.jl#L358 So it will generate an intermediate array and will use ForwardDiff when differentiating (as broadcasting operations with Tracker involve ForwardDiff).